packages/transformers/js/src/visitors/fs.js

```diff
@@ -42,65 +43,125 @@ export default ({
-          .get('arguments')
-          .map(arg => evaluate(arg, vars)): Array<string>);
-
-        filename = Path.resolve(filename);
-        res = fs.readFileSync(filename, ...args);
-      } catch (_err) {
-        let err: Error = _err;
-
-        if (err instanceof NodeNotEvaluatedError) {
-          // Warn using a code frame
-          err.fileName = asset.filePath;
-
-          // $FlowFixMe the actual stack is useless
-          delete err.stack;
-
-          logger.warn(errorToDiagnostic(err));
-          return;
+      let filename, args, res;
+      if (asset.env.unsafeInlining) {
+        try {
+          [filename, ...args] = (path
+            .get('arguments')
+            .map(arg => evaluate(arg, vars)): Array<string>);
+
+          filename = Path.resolve(filename);
+          if (!Path.relative(options.projectRoot, filename).startsWith('..')) {
+            res = options.inputFS.readFileSync(filename, ...args);
+          }
+        } catch (_err) {
+          let err: Error = _err;
+
+          if (err instanceof NodeNotEvaluatedError) {
+            // Warn using a code frame
+            err.fileName = asset.filePath;
+
+            // $FlowFixMe the actual stack is useless
+            delete err.stack;
+
+            logger.warn(errorToDiagnostic(err));
+            return;
+          }
+
+          // Add location info so we log a code frame with the error
+          // $FlowFixMe Added in Flow 0.121.0 upgrade in #4381
+          err.loc =
+            path.node.arguments.length > 0
+              ? path.node.arguments[0].loc?.start
+              : path.node.loc?.start;
+          throw err;
         }
-
-        // Add location info so we log a code frame with the error
-        // $FlowFixMe Added in Flow 0.121.0 upgrade in #4381
-        err.loc =
-          path.node.arguments.length > 0
-            ? path.node.arguments[0].loc?.start
-            : path.node.loc?.start;
-        throw err;
       }
 
       let replacementNode;
-      if (Buffer.isBuffer(res)) {
-        invariant(res != null);
-        replacementNode = bufferTemplate({
-          CONTENT: t.stringLiteral(res.toString('base64')),
+      if (filename && res) {
+        if (Buffer.isBuffer(res)) {
+          invariant(res != null);
+          replacementNode = bufferTemplate({
+            CONTENT: t.stringLiteral(res.toString('base64')),
+          });
+        } else {
+          invariant(typeof res === 'string');
+          replacementNode = t.stringLiteral(res);
+        }
+
+        invariant(filename != null);
+        asset.addIncludedFile({
+          filePath: filename,
         });
       } else {
-        invariant(typeof res === 'string');
-        replacementNode = t.stringLiteral(res);
-      }
+        let loc = convertBabelLoc(path.node.loc);
+        if (filename) {
+          logger.warn({
+            message:
+              'Disallowing fs.readFileSync of file outside project root, replacing with dummy value',
```

Comment:
Should this be an error?

Reply:
You mean inserting `new Error()` or throwing a build error?
https://unpkg.com/browse/is-docker@2.1.1/index.js

Reply:
I was thinking it would fail the build. Inserting some fake content might be weird?